File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/ipc/MessagePortIpc.kt
package org.dweb_browser.core.ipc

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch
import org.dweb_browser.core.help.types.IMicroModuleManifest
import org.dweb_browser.core.ipc.helper.DWebMessage
import org.dweb_browser.core.ipc.helper.DWebMessageBytesEncode
import org.dweb_browser.core.ipc.helper.IWebMessagePort
import org.dweb_browser.core.ipc.helper.IpcMessage
import org.dweb_browser.core.ipc.helper.IpcPoolMessageArgs
import org.dweb_browser.core.ipc.helper.IpcPoolPack
import org.dweb_browser.core.ipc.helper.PackIpcMessage
import org.dweb_browser.core.ipc.helper.cborToIpcMessage
import org.dweb_browser.core.ipc.helper.cborToIpcPoolPack
import org.dweb_browser.core.ipc.helper.ipcMessageToCbor
import org.dweb_browser.core.ipc.helper.ipcPoolPackToCbor
import org.dweb_browser.core.ipc.helper.ipcPoolPackToJson
import org.dweb_browser.core.ipc.helper.jsonToIpcPack
import org.dweb_browser.core.ipc.helper.jsonToIpcPoolPack
import org.dweb_browser.helper.Debugger
import org.dweb_browser.helper.WeakHashMap
import org.dweb_browser.helper.getOrPut

val debugMessagePortIpc = Debugger("message-port-ipc")

class MessagePort(private val port: IWebMessagePort) {
  companion object {
    private val wm = WeakHashMap<IWebMessagePort, MessagePort>()
    suspend fun from(port: IWebMessagePort): MessagePort =
      wm.getOrPut(port) { MessagePort(port).also { it.installMessageSignal() } }
  }

  val webMessageFlow = MutableSharedFlow<DWebMessage>()

  private suspend fun installMessageSignal() {
    port.onMessage {
      webMessageFlow.emit(it)
    }
  }


  suspend fun postMessage(data: String) {
    port.postMessage(DWebMessage.DWebMessageString(data))
  }

  suspend fun postMessage(data: ByteArray) {
    port.postMessage(DWebMessage.DWebMessageBytes(data))
  }

  private var _isClosed = false
  suspend fun close() {
    if (_isClosed) {
      return
    }
    _isClosed = true
    port.close()
  }
}


open class MessagePortIpc(
  val port: MessagePort,
  override val remote: IMicroModuleManifest,
  channelId: String,
  endpoint: IpcPool
) : Ipc(channelId, endpoint) {
  companion object {
    suspend fun from(
      port: IWebMessagePort,
      remote: IMicroModuleManifest,
      channelId: String,
      endpoint: IpcPool
    ) = MessagePortIpc(MessagePort.from(port), remote, channelId, endpoint)

  }

  override fun toString(): String {
    return super.toString() + "@MessagePortIpc($channelId)"
  }

  init {
    ipcScope.launch {
      port.webMessageFlow.collect { event ->
        val ipc = this@MessagePortIpc

        val packMessage = when (event) {
          is DWebMessage.DWebMessageString -> {
            val pack = jsonToIpcPack(event.data)
            val message = jsonToIpcPoolPack(pack.ipcMessage, this@MessagePortIpc)
            debugMessagePortIpc("ON-MESSAGE string") { "$channelId => $message" }
            IpcPoolPack(pack.pid, message)
          }

          is DWebMessage.DWebMessageBytes -> {
            when (event.encode) {
              DWebMessageBytesEncode.Normal -> {
                val pack = jsonToIpcPack(event.data.decodeToString())
                val message = jsonToIpcPoolPack(pack.ipcMessage, ipc)
                debugMessagePortIpc("ON-MESSAGE json") { "$endpoint => $message" }
                IpcPoolPack(pack.pid, message)
              }

              DWebMessageBytesEncode.Cbor -> {
                val pack = cborToIpcPoolPack(event.data)
                val message = cborToIpcMessage(pack.messageByteArray, this@MessagePortIpc)
                debugMessagePortIpc("ON-MESSAGE cbor") { "$endpoint => $message" }
                IpcPoolPack(pack.pid, message)
              }
            }
          }
        }

        endpoint.dispatchMessage(IpcPoolMessageArgs(packMessage, ipc))
      }
    }
  }

  override suspend fun doPostMessage(pid: Int, data: IpcMessage) {
    if (supportCbor) {
      val message = ipcMessageToCbor(data)
      val pack = ipcPoolPackToCbor(PackIpcMessage(pid, message))
      this.port.postMessage(pack)
      return
    }
    // 普通信号
    val message = ipcPoolPackToJson(IpcPoolPack(pid, data))
    try {
      this.port.postMessage(message)
    } catch (e: Exception) {
      println("xpostMessage=> $channelId $data ${e.message}")
    }
  }

  override suspend fun _doClose() {
    this.port.close()
  }

}

File: desktop-dev/src/core/ipc/MessagePortIpc.ts
import { encode } from "cbor-x";

import { IpcPool, IpcPoolPack, IpcPoolPackString } from "../index.ts";
import type { $IpcSupportProtocols, $MicroModuleManifest } from "../types.ts";
import { IpcRequest } from "./ipc-message/IpcRequest.ts";
import { IpcResponse } from "./ipc-message/IpcResponse.ts";
import {
  $cborToIpcMessage,
  $jsonToIpcMessage,
  $messageToIpcMessage,
  $uint8ArrayToIpcMessage,
} from "./helper/$messageToIpcMessage.ts";
import { $IpcMessage } from "./helper/const.ts";
import { Ipc } from "./ipc.ts";

export class MessagePortIpc extends Ipc {
  constructor(
    readonly port: MessagePort,
    readonly remote: $MicroModuleManifest,
    override channelId: string,
    override endpoint: IpcPool,
    readonly self_support_protocols: $IpcSupportProtocols = {
      raw: true,
      cbor: true,
      protobuf: false,
    }
  ) {
    super(channelId, endpoint);

    /** messageport内置JS对象解码，但也要看对方是否支持接受，比如Android层就只能接受String类型的数据 */
    this._support_raw = self_support_protocols.raw && this.remote.ipc_support_protocols.raw;
    /** JS 环境里支持 cbor 协议，但也要看对等方是否支持 */
    this._support_cbor = self_support_protocols.cbor && this.remote.ipc_support_protocols.cbor;

    port.addEventListener("message", (event) => {
      const message = this.support_raw
        ? $messageToIpcMessage(event.data, this)
        : event.data instanceof Uint8Array
        ? $uint8ArrayToIpcMessage(event.data, this)
        : this.support_cbor
        ? $cborToIpcMessage(event.data, this)
        : $jsonToIpcMessage(event.data, this);
      if (message === undefined) {
        console.error("MessagePortIpc.cts unkonwn message", event.data);
        return;
      }
      this.endpoint.emitMessage(message, this);
    });
    port.start();
  }

  _doPostMessage(pid: number, message: $IpcMessage): void {
    // deno-lint-ignore no-explicit-any
    let message_raw: any;
    if (message instanceof IpcRequest) {
      message_raw = message.ipcReqMessage();
    } else if (message instanceof IpcResponse) {
      message_raw = message.ipcResMessage();
    } else {
      message_raw = message;
    }
    const pack = new IpcPoolPack(pid, message_raw);
    // deno-lint-ignore no-explicit-any
    let message_data: any;
    if (this.support_raw) {
      // 内存直传
      message_data = pack;
    } else if (this.support_cbor) {
      message_data = encode(pack);
    } else {
      message_data = JSON.stringify(new IpcPoolPackString(pid, JSON.stringify(message_raw)));
    }
    this.port.postMessage(message_data);
  }

  _doClose() {
    this.port.close();
  }
}
